File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/core/DemoSessionPairer.kt
package com.flagshipserver.app.core

import com.flagshipserver.app.api.DemoServerBlock
import com.flagshipserver.app.api.DemoSessionRecord
import com.flagshipserver.app.api.FlagshipServerClient
import com.flagshipserver.app.api.SessionStoring

/** Passwordless demo profiles hold no owner IRK. The Worker therefore mints
 * their narrowly-scoped paired session, while the native protected store keeps
 * the returned token across process death. */
object DemoSessionPairer {
    suspend fun ensurePaired(
        username: String,
        server: DemoServerBlock,
        client: FlagshipServerClient,
        store: SessionStoring,
        replacingExistingToken: Boolean = false,
        makeToken: () -> String = { AddPairedSessionOrder.freshToken() },
    ): String {
        val podId = PodInfo.podId(server.fqdn)
        val existing = store.sessionToken(forPodId = podId)
        if (!replacingExistingToken && !existing.isNullOrEmpty()) {
            store.activatePod(podId, "https://${server.fqdn}")
            store.setDemoSession(DemoSessionRecord(username, server))
            return existing
        }

        val token = makeToken().lowercase()
        val response = client.pairDemoSession(username, token)
        check(response.ok && response.fqdn.equals(server.fqdn, ignoreCase = true)) {
            "demo pairing returned a different server"
        }
        store.setSessionToken(token, forPodId = podId)
        store.activatePod(podId, "https://${server.fqdn}")
        store.setDemoSession(DemoSessionRecord(username, server))
        return token
    }
}

File: apps/mobile/shared/Sources/FlagshipCore/DemoSessionPairer.swift
import Foundation
import FlagshipAPI

/// Pairs a passwordless demo profile through the Worker's narrow demo signer.
/// Demo identities deliberately have no device-held IRK, so the normal
/// owner-signing path cannot work for them.
public enum DemoSessionPairer {
    @discardableResult
    public static func ensurePaired(
        username: String,
        server: DemoServerBlock,
        client: any FlagshipServerClient,
        store: any SessionStoring,
        replacingExistingToken: Bool = false,
        makeToken: () -> String = { AddPairedSessionOrder.freshToken() }
    ) async throws -> String {
        let podId = PodInfo.podId(forFqdn: server.fqdn)
        if !replacingExistingToken,
           let existing = await store.sessionToken(forPodId: podId), !existing.isEmpty {
            await store.activatePod(podId, baseUrl: "https://\(server.fqdn)")
            await store.setDemoSession(DemoSessionRecord(username: username, server: server))
            return existing
        }

        let token = makeToken().lowercased()
        let response = try await client.pairDemoSession(username: username, token: token)
        guard response.ok, response.fqdn.lowercased() == server.fqdn.lowercased() else {
            throw ScreensClientError.decoding("demo pairing returned a different server")
        }

        await store.setSessionToken(token, forPodId: podId)
        await store.activatePod(podId, baseUrl: "https://\(server.fqdn)")
        await store.setDemoSession(DemoSessionRecord(username: username, server: server))
        return token
    }
}
